server/src/main/java/io/deephaven/server/session/SessionService.java

```diff
@@ -333,21 +338,41 @@ public SessionState getSessionForAuthToken(final String token) throws Authentica
                 if (session != null) {
                     return session;
                 }
-            } catch (IllegalArgumentException ignored) {
+            } catch (IllegalArgumentException | InvalidUuidException ignored) {
             }
+
+            // In case we don't have another handler for Bearer, look for nested tokens to try later
+            int offset = authToken.indexOf(' ');
+            bearerKey = authToken.substring(0, offset < 0 ? authToken.length() : offset);
+            bearerPayload = offset < 0 ? "" : authToken.substring(offset + 1);
         }
 
         int offset = token.indexOf(' ');
         final String key = token.substring(0, offset < 0 ? token.length() : offset);
         final String payload = offset < 0 ? "" : token.substring(offset + 1);
+        // Use the auth type to look up a handler. If this happens to be Bearer, this gives a chance for an auth handler
+        // to claim that type
         AuthenticationRequestHandler handler = authRequestHandlers.get(key);
-        if (handler == null) {
-            log.info().append("No AuthenticationRequestHandler registered for type ").append(key).endl();
-            throw new AuthenticationException();
+        if (handler != null) {
+            Optional<AuthContext> s = handler.login(payload, SessionServiceGrpcImpl::insertCallHeader);
+            if (s.isPresent()) {
+                return newSession(s.get());
+            }
         }
-        return handler.login(payload, SessionServiceGrpcImpl::insertCallHeader)
-                .map(this::newSession)
-                .orElseThrow(AuthenticationException::new);
+        // If nothing succeeded or errored yet, and we found a key in the bearer value, try that next
+        if (bearerKey != null) {
+            handler = authRequestHandlers.get(bearerKey);
+            if (handler != null) {
+                Optional<AuthContext> s = handler.login(bearerPayload, SessionServiceGrpcImpl::insertCallHeader);
+                if (s.isPresent()) {
+                    return newSession(s.get());
+                }
+            }
+        }
+
+        // No more options, log an error and return
+        log.info().append("No AuthenticationRequestHandler registered for type ").append(key).endl();
```

Comment:
I wonder if we should potentially log out `bearerKey` if `bearerOffset >= 0`? I guess technically, you could have an auth handler who's auth token just happens to have spaces in it, in which case printing out even part of the key would be bad...

Also, this log message does seem technically incorrect; there might be an appropriate handler registered, just that "login" did not succeed.

I wonder if we should collect the `handler.getClass()` in both cases and print that out for additional (non-leaky) info?

Reply:
>I wonder if we should potentially log out bearerKey if bearerOffset >= 0? I guess technically, you could have an auth handler who's auth token just happens to have spaces in it, in which case printing out even part of the key would be bad...

I was leaning towards not logging in case there was a space in the token, yes.

> Also, this log message does seem technically incorrect; there might be an appropriate handler registered, just that "login" did not succeed.

Nit, there might be a handler registered, but it might not be the "appropriate" one. Mostly I was aiming to keep the message the same, but we really have three cases here:
 * return auth context, signaling that authentication was successful
 * throw auth exception, signaling that authentication failed
 * return empty, signaling that this handler is not the correct one to handle the request (or the request is ambiguous in some way that the handler doesn't want to say "this login failed")

If a handler doesn't outright succeed or fail, we want to say "no one claimed this auth attempt" - the handler might be registered with a matching name, but it wasn't the correct handler for this request.

> I wonder if we should collect the handler.getClass() in both cases and print that out for additional (non-leaky) info?

Maybe as debug logging? It can be noisy otherwise when not actively debugging auth issues?